waku/README.md

```diff
@@ -0,0 +1,287 @@
+
+---
+
+### `config.go`
+
+[`config.go`](./config.go) is home to the `Config{}` struct and the declaration of `DefaultConfig`.
+
+`Config{}` is used to initialise the settings of an instantiated `Waku{}`. `waku.New()` creates a new instance of a `Waku{}` and takes a `Config{}` as a parameter, if nil is passed instead of an instance of `Config{}`, `DefaultConfig` is used. 
+
+---
+
+### `mailserver.go`
+
+[`mailserver.go`](./mailserver.go) is home to `MailServer` interface, which is implemented by `mailserver.WakuMailServer{}` found in the package file [`mailserver/mailserver.go`](../mailserver/mailserver.go). `MailServer` represents a mail server, capable of receiving and archiving messages for subsequent delivery to the peers.
+
+Additionally this package is home to `MailServerResponse{}` which represents the response payload sent by the mail-server. `MailServerResponse{}` is ultimately initialised by `CreateMailServerEvent()`, which is tied to the main `Waku{}` via the `Waku.OnP2PRequestCompleted()` function. This is ultimately accessed via the `Peer.Run()` function and is made available outside of the package with the `waku.HandlePeer()` function via `Waku.protocol.Run := waku.HandlePeer`. 
+
+---
+
+## Common
+
+### `bloomfilter.go`
+
+[`bloomfilter.go`](./common/bloomfilter.go) holds a few bloomfilter specific functions.
+
+---
+
+### `const.go`
+
+[`const.go`](./common/const.go), originally a hangover from the [`go-ethereum` `whisperv6/doc.go` package file](https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv6/doc.go) later [refactored](https://github.com/status-im/status-go/pull/1950), is home to the common Waku constants.
+
+#### Notes
+
+Versions also have version specific `const.go` files.
+
+---
+
+### `envelope.go`
+
+[`envelope.go`](./common/envelope.go) is home to the `Evelope{}` and `EnvelopeError{}` structs. `Envelope{}` is used as the data packet in which message data is sent through the Waku network.
+
+`Envelope{}` is accessed via the initialisation function `NewEnvelope()`, which is exclusively consumed by `Message.Wrap()` that prepares a message to be sent via Waku. 
+
+---
+
+### `errors.go`
+
+[`errors.go`](./common/errors.go) holds generic package errors.
+
+---
+
+### `events.go`
+
+[`events.go`](./common/events.go) handles data related to Waku events. This file contains string type `const`s that identify known Waku events.
+
+Additionally, the file contains `EnvelopeEvent{}`, which serves as a representation of events created by envelopes. `EnvelopeEvent{}`s are initialised exclusively within the `waku` package.  
+
+--- 
+
+### `filter.go`
+
+[`filter.go`](./common/filter.go) is home to `Filter{}` which represents a waku filter.
+
+#### Usage
+
+A `status-go` node will install / register filters through RPC calls from a client (eg `status-react`). When the client installs a filter, it will specify 2 things:

```

Comment:
This is a bit more complicated:
it allows to specify a `PoW` ( https://github.com/status-im/status-go/blob/4d00656c41909ccdd80a8a77a0982bd66f74d29e/waku/common/filter.go#L37 ) as well, which is the minimum `PoW` for that filter, a public key for the sender (`Src` https://github.com/status-im/status-go/blob/4d00656c41909ccdd80a8a77a0982bd66f74d29e/waku/common/filter.go#L33 )  and whether we accept `P2P` envelopes on this filter https://github.com/status-im/status-go/blob/4d00656c41909ccdd80a8a77a0982bd66f74d29e/waku/common/filter.go#L38 .
We currently don't use `Src` at all.
I believe we always set `P2P` to `true` as we always want to receive `P2P` envelopes on those filters (only from trusted peers).

And we kind of use .`PoW`:

Whisper forced you to listen to a filter if you wanted to publish to it (for light client nodes I am not sure that makes much sense, but we'd have to look into it a bit deeper).

In `status-react` each user listen to the topic `hash(pk)`, so if I want to send a message to `hash(pk1)` I will need to listen to that topic as well, but I don't really want to receive envelopes. So we listen to that topic, but we set the `PoW` to 1 so that all envelopes are discarded.

Not sure we want to document all of these (we could), but otherwise I would just change `it will specify 2 things` to `it will specify at least 2 things` :)

Reply:
It is tricky to know where the line needs to be drawn on the docs. I want to give the read enough information so that they can understand the purpose but at the same time not be so overwhelmed with information that they are confused.

I will at least change the statement that `it will specify at least 2 things` and then detail the optional fields.

Reply:
Addressed with https://github.com/status-im/status-go/pull/1949/commits/223f4d26937d41b9f7a1acb6336581f1157fb290